cmd/createorupdate/createorupdate.go

```diff
@@ -1,88 +1,176 @@
 package main
 
 import (
+	"fmt"
 	"io/ioutil"
 	"os"
 
 	acsapi "github.com/Azure/acs-engine/pkg/api"
 	"github.com/Azure/acs-engine/pkg/api/osa/vlabs"
 	"github.com/ghodss/yaml"
+	"k8s.io/apimachinery/pkg/util/errors"
 
+	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/config"
 	"github.com/openshift/openshift-azure/pkg/plugin"
+	"github.com/openshift/openshift-azure/pkg/tls"
 )
 
-func createOrUpdate() error {
-	b, err := ioutil.ReadFile("_data/manifest.yaml")
-	if err != nil {
-		return err
-	}
-	var ext *vlabs.OpenShiftCluster
-	err = yaml.Unmarshal(b, &ext)
-	if err != nil {
-		return err
-	}
-	err = ext.Validate()
-	if err != nil {
-		return err
+// createOrUpdate simulates the RP
+func createOrUpdate(oc *vlabs.OpenShiftCluster) (*vlabs.OpenShiftCluster, error) {
+	// instantiate the plugin
+	var p api.Plugin = &plugin.Plugin{}
+
+	// validate the external API manifest
+	errs := p.ValidateExternal(oc)
+	if len(errs) > 0 {
+		return nil, errors.NewAggregate(errs)
 	}
-	cs := ext.AsContainerService()
-	err = plugin.Enrich(cs)
+
+	// convert the external API manifest into the internal API representation
+	cs := acsapi.ConvertVLabsOpenShiftClusterToContainerService(oc)
+
+	// the RP will enrich the internal API representation with data not included
+	// in the original request
+	// TODO(mjudeikis): choose DNS names here
+	err := enrich(cs)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	var oldCs *acsapi.ContainerService
+	// read in the OpenShift config blob if it exists (i.e. we're updating)
 	var configBytes []byte
 	if _, err := os.Stat("_data/config.yaml"); err == nil {
-		// pre-existing config means we're in the update path
 		configBytes, err = ioutil.ReadFile("_data/config.yaml")
 		if err != nil {
-			return err
+			return nil, err
 		}
+	}
 
-		// fake this for now
+	// in the update path, the RP should have access to the previous internal
+	// API representation for comparison.  Fake this for now
+	var oldCs *acsapi.ContainerService
+	if len(configBytes) > 0 {
 		oldCs = cs
 	}
 
-	p, err := plugin.NewPlugin(cs, oldCs, configBytes)
+	// validate the internal API representation (with reference to the previous
+	// internal API representation)
+	errs = p.ValidateInternal(cs, oldCs)
+	if len(errs) > 0 {
+		return nil, errors.NewAggregate(errs)
+	}
+
+	// generate or update the OpenShift config blob
+	configBytes, err = p.GenerateConfig(cs, configBytes)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	err = p.Validate()
+	// persist the OpenShift config blob
+	err = ioutil.WriteFile("_data/config.yaml", configBytes, 0600)
```

Comment:
Should we keep some backups of n-1 configuration? would make debugging and rollbacks little bit easier.

Reply:
We have briefly discussed storing separate blobs under the storage account that holds the config.